### ISSUE TITLE
Make AppVeyor run unit tests

### ISSUE THREAD

**Issue by tidyui** [PiranhaCMS/piranha.core#42]
AppVeyor seemed to fail running the unit tests with the following message: 

    Discovering tests...OK
    %xunit20%\xunit.console.x86 "C:\projects\piranha-core\test\core\Piranha.AspNet.Tests\bin\Debug\netcoreapp1.1\Piranha.AspNet.Tests.dll" -appveyor
    xUnit.net Console Runner (32-bit .NET 4.0.30319.42000)
    System.InvalidOperationException: Unknown test framework: could not find xunit.dll (v1) or xunit.execution.*.dll (v2) in C:\projects\piranha-core\test\core\Piranha.AspNet.Tests\bin\Debug\netcoreapp1.1

I'm not sure why this is happening, have you tried running xunit tests within AppVeyor @TheWolfA2

**Comment by TheWolfA2** [comment]
I've honestly never used AppVeyor before. I'm looking into it, and should have a PR in for you before the end of the day that has at least has AppVeyor running

**Comment by TheWolfA2** [comment]
I have most of an AppVeyor setup working, which is good! I need to go through the results of the test and determine if they are "supposed" to be failing, or if it's a separate issue. Once I've gone through the failing tests and figured everything out, I will make the PR with AppVeyor setup notes

**Comment by tidyui** [comment]
Yeah, I took a look at the "failing" tests. In my opinion a Failing Test should succeed if it fails, it shouldn't actually fail, for example:

    void FailingTest() {
      Assert.NotEqual(4 + 3, 6);
    }

This tests that 4 + 3 isn't equal to 6 and succeeds if the comparison isn't true. This is completely different to writing:

    void FailingTest() {
      Assert.Equal(4 + 3, 6);
    }

**Comment by TheWolfA2** [comment]
I totally agree, if you're expecting a "fail" then the test should pass. Those are simply templates that the Yomen xUnit projector generator made. I removed those yesterday since they were just useless white noise

Billy Wolfington

**Comment by TheWolfA2** [comment]
The main concern I have with the unit tests and AppVeyor (and running unit test projects) is that there are inconsistent unit tests results due to race conditions and what I can only assume are multi-threaded test sequencing issues surrounding `Piranha.App`.  I know this is the root of the problem because `Piranha.Manager.Tests.Areas.Manager.Controllers.PageControllerUnitTest` will sporadically throw `NullReferenceException` in my test setup when it tries to construct dummy pages for testing. The only reason it would be getting a null reference is if `Models.Page<DynamicPage>Create()` cannot find the page type specified, since that is the only instance where `return null` occurs.

Ultimately, I need to find a way to run each unit test class in an isolated fashion, because running entire projects causes those tests to overlap their setup and give inconsistent results due to the global nature of `Piranha.App`. The best solution I can think of this point is to write test scripts, which have a precompiled list of unit test classes to run, and a "master script" will iterate over the projects to run each of the sub scripts. So for example

```
# ~/test/RunTests.ps1
foreach ($testProject in ls core/*.Tests) {
    Push-Location $test

    dotnet restore
    dotnet build
    # This would run ~/test/core/Piranha.Manager.Tests/RunTests.ps1
    & RunTests.ps1
}
```

This feels like a very smelly solution, so I'll see if I can do some other kind of method of running unit tests

**Comment by tidyui** [comment]
Hmmm.. well, I think it's a very interesting problem since web applications by nature are mutlithreaded. The initialization of the global **App** object and its resources has to work in a multithreaded environment so running multithreaded tests "should" work unless it's modifying the same data.

I'll take a look at the manager test you refered to to see how everything's set up!

/Håkan

**Comment by TheWolfA2** [comment]
I think part of the issue may be that `App.PageTypes` is treated as a "read-only" object through out the app, and that after load the page types aren't altered? However, in the tests I've written some tests manipulate the data behind the `App.PageTypes` for different testing scenarios.

Plus, each test in `Piranha.Manager.Tests` is assuming it's running in an isolated environment so if another test is manipulating the assumed state of `App.PageTypes` or any of it's global properties then it would cause non-deterministic output once multiple unit test classes are running.

**Comment by TheWolfA2** [comment]
Appreciate it! I think it must be a misunderstanding of mine in how `App.PageTypes` operates, because I just had `PageControllerUnitTest` give me two different fail counts (7, and 9) which would indicate that it's not just when multiple classes are running. I'll see if I can recreate the issue and verify that though.

Some of the tests are currently failing outside of the race condition issue surrounding App, specifically the *ThrowsException tests. These are failing because they are not throwing exceptions but I am of the opinion they should be (discussed in a separate issue).  To highlight the test names specifically, they are:

* `Publish_NewPageWithFailedPublishReturnsView()`
* `Publish_NewPageWithInvalidPageTypeIdThrowsException` (getting renamed to `ThrowsKeyNotFoundException`)
* `Save_NewPageWithFailedSaveReturnsView`
* `Save_NewPageWithInvalidPageTypeIdThrowsException` (getting renamed to `ThrowsKeyNotFoundException`)
* `UnPublish_NewPageWithFailedPublishReturnsView`
* `UnPublish_NewPageWithInvalidPageTypeIdThrowsException` (getting renamed to `ThrowsKeyNotFoundException`)

**Comment by TheWolfA2** [comment]
Alright, I was finally able to establish at least one ordering where a unit test that should pass does not.  Bear with me.

* `PageControllerUnitTest..ctor` is called for test `Add_ResultsWithNoPageTypesThrowsException()`
* `PageControllerUnitTest..ctor` is called for test `Add_WithValidPageTypeIdGivesCorrectPageEditModel()`
* `pageTypes.Clear();` is run **after** `PageControllerUnitTest..ctor()` for `Add_WithValidPageTypeIdGivesCorrectPageEditModel()`
    * This removes all page types from `App.PageTypes` that `Add_WithValidPageTypeIdGivesCorrectPageEditModel()` is expecting to exist
* `Add_WithValidPageTypeIdgivesCorrectPageEditModel()` eventually tries to grab a page type with an Id that no longer exists throwing a `NullReferenceException` once the object is accessed causing the test to fail

This doesn't happen as much when the unit test is run on the class, or when fewer tests are run, since it's fewer lines of code to serialize, but once more lines of code which alter the state of `App.PageTypes` are added the chances of failure increase, especially since the initialization method `PageControllerUnitTest.InitalizePages()` requires PageTypes to exist.

Ultimately what I'm seeing is a classic case of race conditions and sharing of global resources on an unprotected resource. Maybe this case doesn't happen often because, as you mentioned previously, Piranha is used on a smaller scale so there are fewer concurrent edits and therefore fewer chances of this type of PageTypes race occurring. However, the unit tests will always run in parallel in some capacity, unless we explicitly place each Fact/Theory into a separate collection (which would be a nightmare), or enforce some type of ordering (less awful, but still difficult since some type of priority attribute will be needed on each method).

My thoughts on possible solutions are:
* Remove the `Piranha.App` cache entirely, since this creates a race condition, and performs, in my opinion, an unnecessary level of caching
    * Yes, this increases the number of DB hits when making pages, but I think the relative trade-off of hitting the DB as opposed to risking race conditions is a a worthwhile loss of performance
* Find a mechanism to isolate and enforce serialized unit test running
    * This has a major trade-off in that running unit tests will require a greater amount of time since all tests within a project can't overlap
* Find a way to create an injectible `IApp` interface of some kind?
    * This doesn't seem totally plausible, since the idea of `App` is to cache info, however this would still result in a global resource which the unit tests (and ultimately users) are competing for

**Comment by tidyui** [comment]
I poked around with the App object refactoring the code-base to work without `App.PageTypes` however most manager test fails anyway with a `DbUpdateConcurrencyException`. My guess would be that multiple threads share the same mock db context for some reason destroying each others data.

/Håkan

**Comment by TheWolfA2** [comment]
So that's actually a unit test where I'm specifically throwing an exception with the message "DbUpdateException" to simulate if the call to Save() were to fail. Some of the tests are failing by design because they have unexecutable code regions in some scenarios. Or the method being called has an inconsistent error pattern. Take a look at the errors on my repo (https://github.com/TheWolfA2/piranha.core) to see which methods specifically have failures at this point

Billy Wolfington

**Comment by TheWolfA2** [comment]
I'll take a look at the tests and create a separate branch to work out the issues, because I'm showing that all tests are failing now for `PageControllerUnitTest` due to an initialization error in the Markdown object. I'll work on that and get that figured out

**Comment by tidyui** [comment]
This has been solved in the rewrite & the new msbuild tooling. AppVeyor is now running & all tests are clear.
